src/jdk.jdwp.agent/share/native/libjdwp/threadControl.c

```diff
@@ -501,6 +544,95 @@ removeVThreads(JNIEnv *env)
     }
 }
 
+/*
+ * Free (garbage collect) a vthread node if it is unused. This can only be called when
+ * locks are held so we don't need to worry about other threads changing the state
+ * of the ThreadNode while we are looking at it. We also need to make sure the
+ * ThreadNode is not in use for operations like single stepping or invoking.
+ */
+static void
+freeUnusedVThreadNode(JNIEnv *env, ThreadNode* node)
+{
+    if (gdata->includeVThreads) {
+        return;
+    }
+
+    /*
+     * node->suspendCount requires special handling to see if it triggers having
+     * to keep the node around. It's possible for it to be 0 yet we still need to
+     * keep the node around. Also, it's possbile for it to be non-zero yet we
+     * don't need to keep the node around. More details in the comments below.
+     */
+    if (node->suspendCount == 0) {
+        /*
+         * Normally a suspendCount of 0 does not result in having to keep the
+         * node. However, if the suspendAllCount is not 0, then we do. Otherwise
+         * when the node is recreated later it will end up assuming suspendAllCount
+         * as its suspendCount rather than 0, which would be incorrect. This
+         * mismatch in suspend counts happens when there is a suspendAll in place,
+         * and ThreadReference.resume() is used to decrement the thread's
+         * suspendCount to 0.
+         */
+        if (suspendAllCount > 0) {
+            return;
+        }
+    } else {
+        /*
+         * Although at first it might seem that a non-zero suspendCount would require
+         * keeping the node, we don't have too if node->suspendCount == suspendAllCount,
+         * because when the node is recreated it will get suspendAllCount assigned to it.
+         * So we only worry about keeping the node around if the two counts are not equal.
+         */
+        if (node->suspendCount != suspendAllCount) {
+            return;
+        }
+
+    }
+
+    // All of the following conditions must pass if we are to free this node. Note
+    // suspendCount checks were already made above, so are not included below. If
+    // we got here, then suspendCount checks passed w.r.t. being able to free the node.
+    // Also note we don't need to check node->toBeResumed. If it is set, that implies
+    // node->suspendCount > 0, and that will trigger toBeResumed getting set when
+    // the ThreadNode is recreated. See createVThreadNodeIfNeeded().
+    if (!node->suspendOnStart &&
+        node->current_ei == 0 &&
+        node->cleInfo.ei == 0 &&
+        !node->currentInvoke.pending &&
+        !node->currentInvoke.started &&
+        !node->currentInvoke.available &&
+        !node->currentStep.pending &&
+        node->instructionStepMode != JVMTI_ENABLE &&
+        !node->pendingInterrupt &&
+        !node->popFrameEvent &&
+        !node->popFrameProceed &&
+        !node->popFrameThread &&
+        node->pendingStop == NULL)
```

Comment:
It is tricky to verify this list of checks is complete. :)

Reply:
> It is tricky to verify this list of checks is complete. :)

Yes. I did an audit of all the fields. Combined with the testing, I'm pretty certain this is accurate. Eventually I'll be calling this code from more places to be even more aggressive about disposing of ThreadNodes, so that will also help uncover bugs here.